File: dotnet/KclLib.AST/Stmt.cs
// Stmt.cs — Statement hierarchy. Mirrors `ast::Stmt` in `crates/ast/src/ast.rs`.
//
// Rust uses `#[serde(tag = "type")]`, so each variant appears in JSON as
// `{"type": "<Variant>", ...}`. `StmtLoader.StmtFromWire` dispatches on that tag.

using System.Text.Json;

namespace KclLib.AST;

public record ExprStmt(string Type, List<NodeRef<object>>? Exprs = null);

public record UnificationStmt(string Type, object? Target = null, NodeRef<SchemaConfig>? Value = null);

public record AssignStmt(string Type, List<NodeRef<Target>>? Targets = null, NodeRef<object>? Ty = null, NodeRef<object>? Value = null);

public record SchemaStmt(string Type,
    NodeRef<string>? Doc = null,
    NodeRef<string>? Name = null,
    object? ParentName = null,
    object? ForHostName = null,
    bool IsMixin = false,
    bool IsProtocol = false,
    NodeRef<Arguments>? Args = null,
    List<object>? Mixins = null,
    List<NodeRef<object>>? Body = null,
    List<NodeRef<Decorator>>? Decorators = null,
    List<NodeRef<object>>? Checks = null,
    object? IndexSignature = null);

public record SchemaAttr(string Type,
    string Doc = "",
    NodeRef<string>? Name = null,
    string? Op = null,
    NodeRef<object>? Value = null,
    bool IsOptional = false,
    List<NodeRef<Decorator>>? Decorators = null,
    NodeRef<object>? Ty = null);

public record RuleStmt(string Type,
    NodeRef<string>? Doc = null,
    NodeRef<string>? Name = null,
    List<object>? ParentRules = null,
    List<NodeRef<Decorator>>? Decorators = null,
    List<NodeRef<object>>? Checks = null,
    NodeRef<Arguments>? Args = null,
    object? ForHostName = null);

public record ImportStmt(string Type, string? Path = null, string? AsName = null, string? PkgName = null, string? PkgRoot = null);

internal static class StmtLoader
{
    public static object? StmtFromWire(JsonElement el)
    {
        if (el.ValueKind != JsonValueKind.Object) return null;
        var variant = WireHelpers.OptString(el, "type");
        if (variant == null) return null;
        return variant switch
        {
            "Expr" => new ExprStmt("Expr", WireHelpers.NodeListFromWire<object>(el.GetProperty("exprs"), Loaders.ExprFromWire)),
            "Unification" => new UnificationStmt(
                "Unification",
                Target: WireHelpers.OptList<object>(el, "target"),
                Value: WireHelpers.NodeFromWire<SchemaConfig>(el.GetProperty("value"), DtoLoader.SchemaConfigFromWire!)),
            "Assign" => new AssignStmt(
                "Assign",
                Targets: WireHelpers.NodeListFromWire<Target>(el.GetProperty("targets"), DtoLoader.TargetFromWire!),
                Ty: WireHelpers.NodeFromWire<object>(el.GetProperty("ty"), Loaders.TypeFromWire),
                Value: WireHelpers.NodeFromWire<object>(el.GetProperty("value"), Loaders.ExprFromWire)),
            "Schema" => SchemaStmtFromWire(el),
            "SchemaAttr" => SchemaAttrFromWire(el),
            "Rule" => RuleStmtFromWire(el),
            "Import" => ImportStmtFromWire(el),
            _ => new BasicType(variant),
        };
    }

    private static SchemaStmt SchemaStmtFromWire(JsonElement o) => new(
        Type: "Schema",
        Doc: WireHelpers.NodeFromWire<string>(o.GetProperty("doc"), x => x.GetString()!),
        Name: WireHelpers.NodeFromWire<string>(o.GetProperty("name"), x => x.GetString()!),
        ParentName: WireHelpers.OptList<object>(o, "parent_name"),
        ForHostName: WireHelpers.OptList<object>(o, "for_host_name"),
        IsMixin: WireHelpers.OptBool(o, "is_mixin"),
        IsProtocol: WireHelpers.OptBool(o, "is_protocol"),
        Args: WireHelpers.NodeFromWire<Arguments>(o.GetProperty("args"), DtoLoader.ArgumentsFromWire!),
        Mixins: WireHelpers.OptList<object>(o, "mixins"),
        Body: WireHelpers.NodeListFromWire<object>(o.GetProperty("body"), StmtFromWire),
        Decorators: WireHelpers.NodeListFromWire<Decorator>(o.GetProperty("decorators"), DtoLoader.DecoratorFromWire!),
        Checks: WireHelpers.NodeListFromWire<object>(o.GetProperty("checks"), Loaders.ExprFromWire),
        IndexSignature: WireHelpers.OptList<object>(o, "index_signature")
    );

    private static SchemaAttr SchemaAttrFromWire(JsonElement o) => new(
        Type: "SchemaAttr",
        Doc: WireHelpers.OptString(o, "doc") ?? "",
        Name: WireHelpers.NodeFromWire<string>(o.GetProperty("name"), x => x.GetString()!),
        Op: WireHelpers.OptString(o, "op"),
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        IsOptional: WireHelpers.OptBool(o, "is_optional"),
        Decorators: WireHelpers.NodeListFromWire<Decorator>(o.GetProperty("decorators"), DtoLoader.DecoratorFromWire!),
        Ty: WireHelpers.NodeFromWire<object>(o.GetProperty("ty"), Loaders.TypeFromWire)
    );

    private static RuleStmt RuleStmtFromWire(JsonElement o) => new(
        Type: "Rule",
        Doc: WireHelpers.NodeFromWire<string>(o.GetProperty("doc"), x => x.GetString()!),
        Name: WireHelpers.NodeFromWire<string>(o.GetProperty("name"), x => x.GetString()!),
        ParentRules: WireHelpers.OptList<object>(o, "parent_rules"),
        Decorators: WireHelpers.NodeListFromWire<Decorator>(o.GetProperty("decorators"), DtoLoader.DecoratorFromWire!),
        Checks: WireHelpers.NodeListFromWire<object>(o.GetProperty("checks"), Loaders.ExprFromWire),
        Args: WireHelpers.NodeFromWire<Arguments>(o.GetProperty("args"), DtoLoader.ArgumentsFromWire!),
        ForHostName: WireHelpers.OptList<object>(o, "for_host_name")
    );

    private static ImportStmt ImportStmtFromWire(JsonElement o)
    {
        var hasNode = o.TryGetProperty("node", out var nodeEl) && nodeEl.ValueKind == JsonValueKind.Object;
        var node = hasNode ? nodeEl : default;
        return new ImportStmt(
            Type: "Import",
            Path: hasNode ? WireHelpers.OptString(node, "path") : null,
            AsName: hasNode ? WireHelpers.OptString(node, "as_name") : null,
            PkgName: hasNode ? WireHelpers.OptString(node, "pkg_name") : null,
            PkgRoot: hasNode ? WireHelpers.OptString(node, "pkg_root") : null
        );
    }
}

File: wasm/src/ast/_stmt.ts
// _stmt.ts — Statement hierarchy. Mirrors `ast::Stmt` in `crates/ast/src/ast.rs`.
//
// Rust uses `#[serde(tag = "type")]`, so each variant appears in JSON as
// `{"type": "<Variant>", ...}`. `stmtFromWire` dispatches on that tag.

import { nodeFromWire } from "./_base";
import * as _dto from "./_dto";
import * as _expr from "./_expr";
import * as _types from "./_types";

function exprStmtFromWire(w: Record<string, unknown>): unknown {
  return {
    exprs: ((w.exprs as unknown[]) || []).map((e) => nodeFromWire(e as never, _expr.exprFromWire as never)),
  };
}

function unificationStmtFromWire(w: Record<string, unknown>): unknown {
  return {
    target: nodeFromWire(w.target as never, ((x: unknown) => x) as never),
    value: nodeFromWire(w.value as never, _dto.schemaConfigFromWire as never),
  };
}

function assignStmtFromWire(w: Record<string, unknown>): unknown {
  return {
    targets: ((w.targets as unknown[]) || []).map((t) => nodeFromWire(t as never, _dto.targetFromWire as never)),
    ty: nodeFromWire(w.ty as never, _types.typeFromWire as never),
    value: nodeFromWire(w.value as never, _expr.exprFromWire as never),
  };
}

function schemaStmtFromWire(w: Record<string, unknown>): unknown {
  return {
    doc: nodeFromWire(w.doc as never, (x: string) => x),
    name: nodeFromWire(w.name as never, (x: string) => x),
    parentName: nodeFromWire(w.parent_name as never, ((x: unknown) => x) as never),
    forHostName: nodeFromWire(w.for_host_name as never, ((x: unknown) => x) as never),
    isMixin: w.is_mixin === true,
    isProtocol: w.is_protocol === true,
    args: nodeFromWire(w.args as never, _dto.argumentsFromWire as never),
    mixins: ((w.mixins as unknown[]) || []).map((m) => nodeFromWire(m as never, ((x: unknown) => x) as never)),
    body: ((w.body as unknown[]) || []).map((b) => nodeFromWire(b as never, stmtFromWire as never)),
    decorators: ((w.decorators as unknown[]) || []).map((deco) => nodeFromWire(deco as never, _dto.decoratorFromWire as never)),
    checks: ((w.checks as unknown[]) || []).map((c) => nodeFromWire(c as never, _expr.exprFromWire as never)),
    indexSignature: nodeFromWire(w.index_signature as never, ((x: unknown) => x) as never),
  };
}

function schemaAttrFromWire(w: Record<string, unknown>): unknown {
  return {
    doc: w.doc || "",
    name: nodeFromWire(w.name as never, (x: string) => x),
    op: w.op,
    value: nodeFromWire(w.value as never, _expr.exprFromWire as never),
    isOptional: w.is_optional === true,
    decorators: ((w.decorators as unknown[]) || []).map((deco) => nodeFromWire(deco as never, _dto.decoratorFromWire as never)),
    ty: nodeFromWire(w.ty as never, _types.typeFromWire as never),
  };
}

function ruleStmtFromWire(w: Record<string, unknown>): unknown {
  return {
    doc: nodeFromWire(w.doc as never, (x: string) => x),
    name: nodeFromWire(w.name as never, (x: string) => x),
    parentRules: ((w.parent_rules as unknown[]) || []).map((p) => nodeFromWire(p as never, ((x: unknown) => x) as never)),
    decorators: ((w.decorators as unknown[]) || []).map((deco) => nodeFromWire(deco as never, _dto.decoratorFromWire as never)),
    checks: ((w.checks as unknown[]) || []).map((c) => nodeFromWire(c as never, _expr.exprFromWire as never)),
    args: nodeFromWire(w.args as never, _dto.argumentsFromWire as never),
    forHostName: nodeFromWire(w.for_host_name as never, ((x: unknown) => x) as never),
  };
}

function importStmtFromWire(w: Record<string, unknown>): unknown {
  const node = (w.node as Record<string, unknown>) || {};
  return {
    path: node.path,
    asName: node.as_name,
    pkgName: node.pkg_name,
    pkgRoot: node.pkg_root,
  };
}

const REGISTRY: Record<string, (w: Record<string, unknown>) => unknown> = {
  Expr: exprStmtFromWire,
  Unification: unificationStmtFromWire,
  Assign: assignStmtFromWire,
  Schema: schemaStmtFromWire,
  SchemaAttr: schemaAttrFromWire,
  Rule: ruleStmtFromWire,
  Import: importStmtFromWire,
};

export function stmtFromWire(w: Record<string, unknown> | undefined | null): unknown {
  if (!w) return undefined;
  if (!w.type) return undefined;
  const variant = w.type as string;
  const loader = REGISTRY[variant];
  if (loader) return Object.assign({ type: variant }, loader(w));
  return { type: variant };
}